fix "sticky" volume_up button on devices

minlexx opened the pull request:
Electrical pin properties configuration was missing, leading to incorrect bias/pull and possibly input/output mode.

Before fix:
```
# grep gpio7 /sys/kernel/debug/gpio
gpiochip2: GPIOs 639-650, parent: platform/800f000.spmi:pmic@2:gpio@c000, 800f000.spmi:pmic@2:gpio@c000:
 gpio7 : in   high normal  vin-0 no pull                     push-pull  no      atest-1 dtest-0
```

After fix:
```
 gpio7 : in   high normal  vin-0 pull-up 30uA                push-pull  no      atest-1 dtest-0
```

:anger:

minlexx commented:
The same needs to be done for other devices.

The same needs to be done for u-boot also... :see_no_evil:

minlexx commented:
Of course volume_up starts working fine in U-Boot after manually setting `pm660l` GPIO pin7 to `bias-pull-up`, like this
```
pmic dev pmic@2
pmic write 0xc642 0x00
```
writing `PMIC_GPIO_PULL_UP_30` (0x00) to gpio7 (base 0xc000 + pin 0x600) register `REG_DIG_PULL_CTL` (0x42) 🙈

M0Rf30 commented:
This is the output on tulip, without changes:
```
xiaomi-tulip:~$ sudo grep gpio7 /sys/kernel/debug/gpio
 gpio7   : out low  func2 2mA no pull
 gpio70  : in  low  func0 2mA pull down
 gpio71  : in  low  func0 2mA pull down
 gpio72  : in  low  func0 2mA no pull
 gpio73  : in  low  func0 2mA pull down
 gpio74  : in  low  func0 2mA pull down
 gpio75  : in  high func0 2mA no pull
 gpio76  : in  high func0 2mA pull down
 gpio77  : in  low  func0 2mA pull down
 gpio78  : out low  func0 2mA pull down
 gpio79  : in  low  func0 2mA pull down
 gpio7 : out  low  func1   vin-0 pull-down 10uA              open-source medium  atest-1 dtest-0
 gpio7 : in   high normal  vin-0 pull-up 30uA                push-pull  no      atest-1 dtest-0
 ```
 Seems that this commit is not required for tulip if I understand correctly.

minlexx commented:
It is obviously needed on my devices, as somehow pmic@2 (pm660l) gpio7 is not configured properly and volume_up button is simply not working, even in Linux. Withouth the fix, no-pull is set. Tested using `libinput debug-events`, button press event arrives instantly, but button release comes only after 10 seconds on lavender and only after 32 seconds on clover. I've measured several times, with stopwatch.

Perhaps it works by lucky coincidence, maybe primary BL configures it correctly, I dunno how I get to this state, maybe u-boot messes it up? But anyway it's always good idea to describe pinconf/pinmux of the pin that you're going to use in GPIO mode.

U-boot fixes are of course required, will come later in u-boot repo..

minlexx commented:
U-Boot part - https://github.com/sdm660-mainline/u-boot/pull/5

minlexx committed:
fixup! arm64: dts: qcom: sdm630: Add initial device tree for nokia-pl2

minlexx committed:
fixup! arm64: dts: qcom: add device-tree for Redmi Note 6 Pro (tulip)

minlexx committed:
fixup! arm64: dts: qcom: sdm660: Add device tree for Xiaomi Mi A2 (jasmine)

minlexx committed:
fixup! arm64: dts: qcom: sdm660-xiaomi-clover: Split to common base and device DTS

minlexx committed:
arm64: dts: qcom: sdm660-lavender: Add pinconf for vol_down gpio

Correct pin configuration for pm660l gpio 7 is missing, which leads
to volume up key behaving werdly sometimes, not registering button up
event (gpio transition to high state) for up to 32 seconds. It needs
to be pulled up to work correctly.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>
